File: scala/src/main/scala/functional/game.scala
package functional

  import scala.annotation.tailrec
  import scala.io.StdIn

object game {

  import functions._
  import GameStatus._

  def main(args: Array[String]): Unit = {

    implicit val board: Board = buildBoard(5)
    printStatus
    val firstHoleIndex = StdIn.readLine("Escolha um buraco para iniciar: ").toInt
    val gameResult = play(firstHoleIndex)
    gameResult match {
      case Won => println("Você ganhou!")
      case Lost => println("Você perdeu!")
    }
  }

  def printStatus(implicit board: Board): Unit = println(board.getStatus)

  def printStatus(currentHoleIndex: Int)(implicit board: Board): Unit = println(board.getStatus(currentHoleIndex))

  @tailrec
  def play(holeIndex: Int)(implicit board: Board): GameStatus = {
    val (updatedBoard, finalHoleIndex) = harvestAndSeedAll(board, holeIndex)
    printStatus(finalHoleIndex)(updatedBoard)
    checkWinningCondition(finalHoleIndex)(updatedBoard) match {
      case Ongoing =>
        val nextHoleIndex = updatedBoard.holes(finalHoleIndex) match {
          case _: Hole => finalHoleIndex
          case _: Rumba => StdIn.readLine("Escolha o próximo buraco: ").toInt
        }
        play(nextHoleIndex)(updatedBoard)
      case default => default
    }
  }

  def checkWinningCondition(currentHoleIndex: Int)(implicit board: Board): GameStatus = {
    if (isRumbaFull(board)) Won
    else if (currentHoleIndex > 0 && board.holes(currentHoleIndex).population == 1) Lost
    else Ongoing
  }

}

object GameStatus extends Enumeration {
  type GameStatus = Value
  val Won, Lost, Ongoing = Value
}

File: naive_python/game.py
from board import Board


def play(board, hole_index):
    last_hole_index = board.harvest_and_seed(hole_index)
    if is_rumba(last_hole_index):
        if check_win_condition(board):
            return "You won!"
        else:
            hole_index = int(input("Please input a hole_index"))
            play(board, hole_index)
    elif was_empty_hole(board.holes[hole_index]):
        return "Game over! You lost!"
    else:
        play(board, last_hole_index)


def is_rumba(hole_index):
    return hole_index == 0


def check_win_condition(board):
    total_seeds = 2 * (len(board.holes) - 1)
    return board.holes[0].population == total_seeds


def was_empty_hole(hole):
    return hole.population == 1


if __name__ == '__main__':
    board = Board(5)
    hole_index = int(input("Please input a hole_index"))
    game_status = play(board, hole_index)
    print(game_status)
